### PR TITLE
Replace errors with warnings for eval args

### PR DESCRIPTION
**Pull request by mvpatel2000**
# What does this PR do?

Replace errors with warnings for eval args. 

Along for the ride, cleans up some of the seq parallel stuff by centralizing validation

### CODE REVIEW

File: composer/core/__init__.py [hunk 1]
```diff
@@ -45,5 +45,5 @@
     'JSON',
     'MemoryFormat',
     'TrainerMode',
-    'validate_eval_automicrobatching',
+    'validate_evaluator',
```
Comment (aspfohl) [inline composer/core/__init__.py:48]: Why is this included as part of composer.core? This is backwards breaking right?

File: composer/core/evaluator.py [hunk 2]
```diff
@@ -141,17 +141,27 @@ def ensure_evaluator(evaluator: Union[Evaluator, DataSpec, Iterable, Dict[str, A
         )
 
 
-def validate_eval_automicrobatching(auto_microbatching: bool, device: Device):
```
Comment (b-chu) [inline composer/core/evaluator.py:144]: Doesn't this break the API?

File: composer/core/evaluator.py [hunk 3]
```diff
@@ -141,17 +141,27 @@ def ensure_evaluator(evaluator: Union[Evaluator, DataSpec, Iterable, Dict[str, A
...
     `_is_auto_microbatching` check because `device` is not available during `Evaluator`
     initialization.
     """
+    auto_microbatching = evaluator.auto_microbatching
     if auto_microbatching and not isinstance(device, DeviceGPU):
         raise ValueError(
             'Can only use adaptive device_eval_microbatch_size on GPU. Please set device_eval_microbatch_size >= 1.',
         )
+    if evaluator.auto_microbatching and hasattr(evaluator.dataloader, 'seq_parallel_world_size'):
+        raise ValueError('Auto microbatching on evaluators is not compatible with sequence parallelism.')
```
Comment (aspfohl) [inline composer/core/evaluator.py:157]: For my understanding: I either need to set a specific evaluator microbatch or turn off sequence parallelism? Could this message be more actionable?

File: composer/core/evaluator.py [hunk 4]
```diff
@@ -141,17 +141,27 @@ def ensure_evaluator(evaluator: Union[Evaluator, DataSpec, Iterable, Dict[str, A
...
+    if isinstance(evaluator.dataloader.get_num_samples_in_batch, int) and hasattr(
+        evaluator.dataloader,
+        'seq_parallel_world_size',
+    ) and evaluator.dataloader.get_num_samples_in_batch * evaluator.dataloader.seq_parallel_world_size != 1:  # type: ignore
+        raise ValueError(
+            '`Sequence parallelism requires a microbatch size of 1 distributed over the sequence parallel group.',
```
Comment (aspfohl) [inline composer/core/evaluator.py:163]:
```suggestion
            'Sequence parallelism requires a microbatch size of 1 distributed over the sequence parallel group.',
```

File: composer/core/evaluator.py [hunk 5]
```diff
@@ -141,17 +141,27 @@ def ensure_evaluator(evaluator: Union[Evaluator, DataSpec, Iterable, Dict[str, A
...
     if auto_microbatching and not isinstance(device, DeviceGPU):
         raise ValueError(
             'Can only use adaptive device_eval_microbatch_size on GPU. Please set device_eval_microbatch_size >= 1.',
         )
+    if evaluator.auto_microbatching and hasattr(evaluator.dataloader, 'seq_parallel_world_size'):
+        raise ValueError('Auto microbatching on evaluators is not compatible with sequence parallelism.')
+    if isinstance(evaluator.dataloader.get_num_samples_in_batch, int) and hasattr(
+        evaluator.dataloader,
+        'seq_parallel_world_size',
+    ) and evaluator.dataloader.get_num_samples_in_batch * evaluator.dataloader.seq_parallel_world_size != 1:  # type: ignore
```
Comment (aspfohl) [inline composer/core/evaluator.py:161]: 👀 what are you ignoring? Should fix

File: composer/core/evaluator.py (hunk 2 above)
Comment (mvpatel2000) [reply composer/core/evaluator.py:144]: Technically yes, but this is mostly an internal validation function -- given we are pre 1.0, I think it's fine to do something like this which cleans up code and almost certainly is not used externally

File: composer/trainer/trainer.py [hunk 6]
```diff
@@ -2116,22 +2107,21 @@ def fit(
...
-                    raise ValueError('Specifying `eval_subset_num_batches` without an `eval_dataloader` has no effect.')
-                if eval_interval != 1:
-                    raise ValueError('Specifying `eval_interval` without an `eval_dataloader` has no effect.')
+                    warnings.warn(
+                        f'Specifying `eval_subset_num_batches={eval_subset_num_batches}` without an `eval_dataloader` '
+                        'has no effect. If trying to run an evaluator, make sure `eval_dataloader` is specified. '
+                        'Otherwise, set `eval_subset_num_batches` to default value -1.',
+                    )
+                if eval_interval != 0 and eval_interval != 1:
+                    warnings.warn(
```
Comment (aspfohl) [inline composer/trainer/trainer.py:2120]: Will this raise multiple warnings?

Comment (mvpatel2000) [reply composer/trainer/trainer.py:2120]: It should raise multiple warnings as the two args are different? Do you want to batch?

File: composer/core/evaluator.py (hunk 5 above)
Comment (mvpatel2000) [reply composer/core/evaluator.py:161]: seq_parallel_world_size is not typed into torch and is outside our control :(

File: composer/core/evaluator.py (hunk 2 above)
Comment (aspfohl) [reply composer/core/evaluator.py:144]: Rather, could we remove this from composer.core if this is the case?

File: composer/core/__init__.py (hunk 1 above)
Comment (mvpatel2000) [reply composer/core/__init__.py:48]: https://github.com/mosaicml/composer/pull/3100#discussion_r1518223176

File: composer/core/evaluator.py (hunk 2 above)
Comment (mvpatel2000) [reply composer/core/evaluator.py:144]: Yea good suggestion. Moved to trainer as a private fn

File: composer/trainer/trainer.py (hunk 6 above)
Comment (aspfohl) [reply composer/trainer/trainer.py:2145]: Ah gotcha. Probably easier to keep as is, batching would just complicate the code

**Review by dakinggg — APPROVED** [review]
